data/fin.lean

```diff
@@ -24,6 +21,85 @@ by cases j; simp [fin.succ]
 @[simp] lemma pred_val (j : fin (n+1)) (h : j ≠ 0) : (j.pred h).val = j.val.pred :=
 by cases j; simp [fin.pred]
 
+@[simp] lemma fin.succ_pred (i : fin (n+1)) (H : i ≠ 0) :
+  (i.pred H).succ = i :=
+begin
+  apply fin.eq_of_veq,
+  cases i with i hi,
+  cases i,
+  { exfalso, apply H, apply fin.eq_of_veq, refl },
+  refl
+end
+
+@[simp] lemma fin.pred_succ (i : fin n) (H : i.succ ≠ 0) :
+  i.succ.pred H = i :=
+by cases i; refl
+
+/-- Embedding of `fin n` in `fin (n+1)` -/
+def raise (k : fin n) : fin (n + 1) := ⟨val k, lt_succ_of_lt (is_lt k)⟩
+
+@[simp] lemma fin.raise_val (k : fin n) : k.raise.val = k.val := rfl
+
+def lower : Π i : fin (n+1), i.1 < n → fin n := λ i h, ⟨i.1, h⟩
+
+@[simp] lemma lower_val (k : fin (n+1)) (H : k.1 < n) : (k.lower H).val = k.val := rfl
+
+def ascend (pivot : fin (n+1)) : Π i : fin n, fin (n+1) :=
+λ i, if i.1 < pivot.1 then i.raise else i.succ
+
+def descend (pivot : fin (n+1)) : Π i : fin (n+1), i ≠ pivot → fin n :=
+λ i H, if h : i.1 < pivot.1
+  then i.lower (lt_of_lt_of_le h $ nat.le_of_lt_succ pivot.2)
+  else i.pred (λ H1, H $ by subst H1;
+    replace h := nat.eq_zero_of_le_zero (le_of_not_gt h);
+    from fin.eq_of_veq h.symm)
+
+theorem ascend_ne (pivot : fin (n+1)) (i : fin n) :
+  pivot.ascend i ≠ pivot :=
+λ H, begin
+  unfold fin.ascend at H,
+  split_ifs at H;
+  rw ← H at h;
+  simp [lt_irrefl, nat.lt_succ_self] at h;
+  exact h
+end
+
+@[simp] lemma ascend_descend (pivot i : fin (n+1))
+  (H : i ≠ pivot) : pivot.ascend (pivot.descend i H) = i :=
+begin
+  unfold fin.descend fin.ascend,
+  split_ifs with H1 H2 H3; apply fin.eq_of_veq; simp at *,
+  { cases pivot with p hp,
+    cases i with i hi,
+    cases i with i, { simp at * },
+    exfalso, apply H, apply fin.eq_of_veq,
+    apply le_antisymm, { apply nat.succ_le_of_lt H2 },
+    simpa using H1 }
+end
+
+@[simp] lemma descend_ascend (pivot : fin (n+1))
+  (i : fin n) (H : pivot.ascend i ≠ pivot) :
+  pivot.descend (pivot.ascend i) H = i :=
+begin
+  unfold fin.descend fin.ascend,
+  apply fin.eq_of_veq,
+  by_cases h : i.val < pivot.val,
+  { simp [h] },
+  { unfold ite dite,
```

Comment:
uh, don't unfold `ite` and `dite`. Use `split_ifs`, or explicit rewrites using `dif_pos` etc...